### PR TITLE
[multistage] Add Multi Stage Strict Replica Group Routing Strategy

### PR DESCRIPTION
**Pull request by ankitsultana**
Multistage queries can involve multiple tables in the same query. To take advantage of tables that have the same instance partitions and are sharded using the join-key, we need a routing strategy that can route to the same replica-group for all the tables involved in the query.

The current routing strategies can't support this because they use IdealState for determining replica-groups, which may differ from InstancePartitions.

**Test Plan**

Tested on local by running a join for two tables with same instancePartitions (replicaGroups=2, numPartitions=2)

```
SELECT
  A.deviceOS,
  A.userUUID
FROM
  userAttributes_OFFLINE AS A
  JOIN userGroups_OFFLINE AS B ON A.userUUID = B.userUUID
  JOIN userGroups_OFFLINE AS C ON B.userUUID = C.userUUID
WHERE
  B.groupUUID = 'group-1'
  AND C.groupUUID = 'group-2'
```

### CODE REVIEW

File: pinot-broker/src/main/java/org/apache/pinot/broker/api/resources/PinotBrokerDebug.java [hunk 1]
```diff
@@ -102,15 +102,15 @@ public Map<String, Map<ServerInstance, List<String>>> getRoutingTable(
     if (tableType != TableType.REALTIME) {
       String offlineTableName = TableNameBuilder.OFFLINE.tableNameWithType(tableName);
       RoutingTable routingTable = _routingManager.getRoutingTable(
-          CalciteSqlCompiler.compileToBrokerRequest("SELECT * FROM " + offlineTableName));
+          CalciteSqlCompiler.compileToBrokerRequest("SELECT * FROM " + offlineTableName), 0);
```
Comment (Jackie-Jiang) [inline pinot-broker/src/main/java/org/apache/pinot/broker/api/resources/PinotBrokerDebug.java:105]: IIRC, presto connector relies on this to return the routing table. Let's add a `requestIdGenerator` in this class to avoid always using the same request id.

File: pinot-broker/src/main/java/org/apache/pinot/broker/routing/instanceselector/BaseInstanceSelector.java [hunk 2]
```diff
@@ -295,6 +289,6 @@ public SelectionResult select(BrokerRequest brokerRequest, List<String> segments
    * <p>NOTE: {@code segmentToEnabledInstancesMap} might contain {@code null} values (segment with no enabled
    * ONLINE/CONSUMING instances). If enabled instances are not {@code null}, they are sorted in alphabetical order.
    */
-  abstract Map<String, String> select(List<String> segments, int requestId,
+  abstract Map<String, String> select(List<String> segments, long requestId,
```
Comment (Jackie-Jiang) [inline pinot-broker/src/main/java/org/apache/pinot/broker/routing/instanceselector/BaseInstanceSelector.java:292]: Let's not change this signature. It expects a non-negative int

File: pinot-broker/src/main/java/org/apache/pinot/broker/routing/instanceselector/BaseInstanceSelector.java [hunk 3]
```diff
@@ -267,8 +262,7 @@ private List<String> calculateEnabledInstancesForSegment(String segment, List<St
   }
 
   @Override
-  public SelectionResult select(BrokerRequest brokerRequest, List<String> segments) {
-    int requestId = (int) (_requestId.getAndIncrement() % MAX_REQUEST_ID);
+  public SelectionResult select(BrokerRequest brokerRequest, List<String> segments, long requestId) {
```
Comment (Jackie-Jiang) [inline pinot-broker/src/main/java/org/apache/pinot/broker/routing/instanceselector/BaseInstanceSelector.java:265]: Keep the `MAX_REQUEST_ID` and pass `(int) (requestId % MAX_REQUEST_ID);`

File: pinot-broker/src/main/java/org/apache/pinot/broker/routing/instanceselector/InstanceSelectorFactory.java [hunk 4]
```diff
@@ -38,7 +40,7 @@ private InstanceSelectorFactory() {
   public static final String LEGACY_REPLICA_GROUP_REALTIME_ROUTING = "PartitionAwareRealtime";
 
   public static InstanceSelector getInstanceSelector(TableConfig tableConfig, BrokerMetrics brokerMetrics,
-      @Nullable AdaptiveServerSelector adaptiveServerSelector) {
+      ZkHelixPropertyStore<ZNRecord> propertyStore, @Nullable AdaptiveServerSelector adaptiveServerSelector) {
```
Comment (Jackie-Jiang) [inline pinot-broker/src/main/java/org/apache/pinot/broker/routing/instanceselector/InstanceSelectorFactory.java:43]: (nit) Put `propertyStore` before `brokerMetrics`

**Review by Jackie-Jiang — COMMENTED** [review]
LGTM in general

File: pinot-broker/src/test/java/org/apache/pinot/broker/routing/instanceselector/InstanceSelectorTest.java [hunk 5]
```diff
@@ -814,14 +837,121 @@ public void testReplicaGroupInstanceSelectorNumReplicaGroupsNotSet() {
...
     for (String segment: segments) {
       expectedReplicaGroupInstanceSelectorResult.put(segment, instance1);
     }
-    selectionResult = replicaGroupInstanceSelector.select(brokerRequest, segments);
+    selectionResult = replicaGroupInstanceSelector.select(brokerRequest, segments, 1);
+    assertEquals(selectionResult.getSegmentToInstanceMap(), expectedReplicaGroupInstanceSelectorResult);
+  }
+
+  @Test
+  public void testMultiStageStrictReplicaGroupSelector() {
```
Comment (ankitsultana) [inline pinot-broker/src/test/java/org/apache/pinot/broker/routing/instanceselector/InstanceSelectorTest.java:851]: self-review: add comments to describe what the test is doing.

File: pinot-broker/src/main/java/org/apache/pinot/broker/api/resources/PinotBrokerDebug.java [hunk 6]
```diff
@@ -157,4 +163,8 @@ public Map<ServerInstance, List<String>> getRoutingTableForQuery(
   public String getServerRoutingStats() {
     return _serverRoutingStatsManager.getServerRoutingStatsStr();
   }
+
+  private int getRequestId() {
+    return (int) (_requestIdGenerator.getAndIncrement() % MAX_REQUEST_ID);
```
Comment (Jackie-Jiang) [inline pinot-broker/src/main/java/org/apache/pinot/broker/api/resources/PinotBrokerDebug.java:168]: I don't think we need to limit the max request id here. It is done in the `InstanceSelector`

Comment (ankitsultana) [reply pinot-broker/src/main/java/org/apache/pinot/broker/api/resources/PinotBrokerDebug.java]: Ah good point. Done.

File: pinot-broker/src/main/java/org/apache/pinot/broker/routing/instanceselector/MultiStageStrictReplicaGroupSelector.java [hunk 7]
```diff
@@ -0,0 +1,150 @@
...
+  public void onAssignmentChange(IdealState idealState, ExternalView externalView, Set<String> onlineSegments) {
+    super.onAssignmentChange(idealState, externalView, onlineSegments);
+    _instancePartitions = getInstancePartitions();
+  }
+
+  @Override
+  Map<String, String> select(List<String> segments, int requestId,
+      Map<String, List<String>> segmentToEnabledInstancesMap, Map<String, String> queryOptions) {
+    // Create a copy of InstancePartitions to avoid race-condition with event-listeners above.
+    InstancePartitions localInstancePartitions = _instancePartitions.createCopy();
```
Comment (Jackie-Jiang) [inline pinot-broker/src/main/java/org/apache/pinot/broker/routing/instanceselector/MultiStageStrictReplicaGroupSelector.java:90]: We don't need to create a copy here. Instead, you may define `_instancePartitions` as `volatile`, and assign it to a local variable
```suggestion
    InstancePartitions instancePartitions = _instancePartitions;
```

File: pinot-common/src/main/java/org/apache/pinot/common/assignment/InstancePartitions.java [hunk 8]
```diff
@@ -134,6 +134,10 @@ public InstancePartitions withName(String newName) {
     return new InstancePartitions(newName, getPartitionToInstancesMap());
   }
 
+  public InstancePartitions createCopy() {
```
Comment (Jackie-Jiang) [inline pinot-common/src/main/java/org/apache/pinot/common/assignment/InstancePartitions.java:137]: This is not needed (see the above comment). Also, this is creating a shallow copy, which actually shares the underlying map

**Review by Jackie-Jiang — APPROVED** [review]
LGTM otherwise

File: pinot-common/src/main/java/org/apache/pinot/common/assignment/InstancePartitions.java (hunk 8 above)
Comment (ankitsultana) [reply pinot-common/src/main/java/org/apache/pinot/common/assignment/InstancePartitions.java]: Ah that was a bad oversight. Thanks for catching this. Fixed.

File: pinot-broker/src/main/java/org/apache/pinot/broker/routing/instanceselector/MultiStageStrictReplicaGroupSelector.java (hunk 7 above)
Comment (ankitsultana) [reply pinot-broker/src/main/java/org/apache/pinot/broker/routing/instanceselector/MultiStageStrictReplicaGroupSelector.java]: Yeah good point. I think in a previous commit I was modifying these but then I changed to a simple reference switch. Fixed.

File: pinot-broker/src/main/java/org/apache/pinot/broker/routing/instanceselector/BaseInstanceSelector.java [hunk 9]
```diff
@@ -50,9 +49,8 @@ abstract class BaseInstanceSelector implements InstanceSelector {
   private static final Logger LOGGER = LoggerFactory.getLogger(BaseInstanceSelector.class);
 
   // To prevent int overflow, reset the request id once it reaches this value
-  private static final int MAX_REQUEST_ID = 1_000_000_000;
+  private static final long MAX_REQUEST_ID = 1_000_000_000;
```
Comment (yupeng9) [inline pinot-broker/src/main/java/org/apache/pinot/broker/routing/instanceselector/BaseInstanceSelector.java:52]: is this change needed?

File: pinot-broker/src/main/java/org/apache/pinot/broker/routing/instanceselector/InstanceSelector.java [hunk 10]
```diff
@@ -56,7 +56,7 @@ public interface InstanceSelector {
    * to selected server instance hosting the segment and a set of unavailable segments (no enabled instance or all
    * enabled instances are in ERROR state).
```
Comment (yupeng9) [inline pinot-broker/src/main/java/org/apache/pinot/broker/routing/instanceselector/InstanceSelector.java:57]: add to java doc about this new parameter

File: pinot-broker/src/main/java/org/apache/pinot/broker/routing/instanceselector/MultiStageStrictReplicaGroupSelector.java [hunk 11]
```diff
@@ -0,0 +1,150 @@
...
+import org.slf4j.LoggerFactory;
+
+
+/**
+ * Instance selector for multi-stage queries which can ensure that Colocated Tables always leverage Colocated Join
+ * whenever possible. To achieve this, this instance-selector uses InstancePartitions (IP) to determine replica-groups,
+ * as opposed to IdealState used by other instance-selectors. Moreover, this also uses the requestId generated by
+ * Pinot broker to determine the replica-group picked for each table involved in the query, as opposed to using a
+ * member variable. There may be scenarios where an instance in the chosen replica-group is down. In that case, this
+ * strategy will try to pick another replica-group. For realtime tables, this strategy uses only CONSUMING partitions.
```
Comment (yupeng9) [inline pinot-broker/src/main/java/org/apache/pinot/broker/routing/instanceselector/MultiStageStrictReplicaGroupSelector.java:50]: if picking another replica-group is fine, then why do we still need the requestID?

File: pinot-broker/src/main/java/org/apache/pinot/broker/routing/instanceselector/BaseInstanceSelector.java (hunk 9 above)
Comment (ankitsultana) [reply pinot-broker/src/main/java/org/apache/pinot/broker/routing/instanceselector/BaseInstanceSelector.java:52]: Technically no, but since I was taking `requestId % MAX_REQUEST_ID` in `select` and `requestId` is a long there thought of explicitly making it a long.

File: pinot-broker/src/main/java/org/apache/pinot/broker/routing/instanceselector/MultiStageStrictReplicaGroupSelector.java (hunk 11 above)
Comment (ankitsultana) [reply pinot-broker/src/main/java/org/apache/pinot/broker/routing/instanceselector/MultiStageStrictReplicaGroupSelector.java:50]: `requestId` is the ID created by the Broker for the entire query. We use that to determine which replica-group to pick for all tables. That ensures that the same replica-group is picked for all tables involved in the join.

In context of colocated joins, we will have multiple tables with the same instance partitions. In the happy path, requestId will be used to select the replica-group and we'll select the same set of servers for each table involved.

When an instance in a replica-group is down or a few segments are in a bad state for a table in one of the replicas, we may try to pick another replica-group. There are two possibilities in this case:

1. We still end up picking the same replica-group for all tables involved in the query. This is because if an instance itself is down, it would impact all tables involved, so ideally the next available replica-group should be picked for all tables. This means the query would still use colocated joins.
2. We pick different replica-groups for some of the tables. This would mean that the query won't use colocated join but it would still be able to run.

I think it may also make sense to instead fail the query since technically we weren't able to guarantee that the same replica-group would be used for the full query. I was thinking of following up on this after running some tests in our internal clusters. Hence I marked this feature as Beta for now.

Comment (yupeng9) [reply pinot-broker/src/main/java/org/apache/pinot/broker/routing/instanceselector/MultiStageStrictReplicaGroupSelector.java:50]: Thanks, your explanation is the same as my understanding. My question is if we shall allow the happy path only, given the the strategy name is `strict replica group`. 

Perhaps we can add another strategy or a flag to allow graceful degadation to normal join.

Comment (ankitsultana) [reply pinot-broker/src/main/java/org/apache/pinot/broker/routing/instanceselector/MultiStageStrictReplicaGroupSelector.java:50]: Yeah I think that makes sense.

One idea I had was to rename this to "Multistage Replica Group" routing since this allows using a different replica-group if one is unavailable, and add another "Multistage Strict Replica Group" routing which would throw an error if a replica-group is down.

I was thinking of trying it out on our clusters for a few days and see how it goes.

In this PR I can also rename this routing strategy to `multiStageReplicaGroup` (from `multiStageStrictReplicaGroup`).

That way we won't need to make a b/w incompatible change on master in the future. This is marked as Beta anyways.

@yupeng9 lmk your thoughts.

Comment (yupeng9) [reply pinot-broker/src/main/java/org/apache/pinot/broker/routing/instanceselector/MultiStageStrictReplicaGroupSelector.java:50]: Sure. Renaming sounds like a good idea, though I'm curious if we still need a default `MultistageReplicaGroup` without the request ID. Or we always want the request ID for multi-stage query execution?

Comment (ankitsultana) [reply pinot-broker/src/main/java/org/apache/pinot/broker/routing/instanceselector/MultiStageStrictReplicaGroupSelector.java]: Renamed.

For multi-stage queries any routing strategy would work. This routing strategy is more focused on ensuring we can leverage colocated joins. Perhaps we should rename this to something that explicitly mentions colocated join in the future.
